### PR TITLE
Add extra loop time to scheduler if not achieving tasks

### PR DESCRIPTION
**Pull request by tridge**
This adds an extra budget of time per loop when we are not achieving scheduled tasks at a rate of at least 1/4 of the desired rate. This fixes an issue where a vehicle can become uncontrollable if the user asks for a SCHED_LOOP_RATE which is not achievable. As these events happen we add extra loop budget until we are able to run all tasks. We drop the extra time when the CPU pressure eases.
This was prompted by a flight by @lthall where the loop rate was set beyond what could be achieved and RC control of the vehicle was lost.
With this change the vehicle should remain controllable no matter how high the desired loop rate it set

### CODE REVIEW

File: libraries/AP_Scheduler/AP_Scheduler.h [hunk 1]
```diff
@@ -200,6 +205,19 @@ class AP_Scheduler
 
     // bitmask bit which indicates if we should log PERF message
     uint32_t _log_performance_bit;
+
+    // maximum task slowdown compared to desired task rate before we
+    // start giving extra time per loop
+    const uint32_t max_task_slowdown = 8;
```
Comment (peterbarker) [inline libraries/AP_Scheduler/AP_Scheduler.h]: uint32_t is somewhat overkill for a `const`.

File: libraries/AP_Scheduler/PerfInfo.cpp [hunk 2]
```diff
@@ -114,13 +115,14 @@ float AP::PerfInfo::get_filtered_time() const
...
-                    "PERF: %u/%u max=%lu min=%lu F=%u sd=%lu",
+                    "PERF: %u/%u [%lu:%lu] F=%u sd=%lu Ex=%lu",
                     (unsigned)get_num_long_running(),
                     (unsigned)get_num_loops(),
                     (unsigned long)get_max_time(),
                     (unsigned long)get_min_time(),
                     (unsigned)(get_filtered_time()*1.0e6),
-                    (unsigned long)get_stddev_time());
+                    (unsigned long)get_stddev_time(),
+                    AP::scheduler().get_extra_loop_us());
```
Comment (peterbarker) [inline libraries/AP_Scheduler/PerfInfo.cpp]: Cast?  This is probably a compiler warning on some platform or another.

File: libraries/AP_Scheduler/AP_Scheduler.h (hunk 1 above)
Comment (tridge) [reply libraries/AP_Scheduler/AP_Scheduler.h]: fixed

File: libraries/AP_Scheduler/PerfInfo.cpp (hunk 2 above)
Comment (tridge) [reply libraries/AP_Scheduler/PerfInfo.cpp]: fixed

**Comment by tridge** [comment]
I flew this on a QAV-250 with SCHED_LOOP_RATE=1200 and it flew fine. It achieved 850Hz in STABILIZE, 650Hz in LOITER

**Comment by WickedShell** [comment]
How does this interact with stuff like the motors library which expect the loop rate as a constant when they are allocated/inited?

**Comment by tridge** [comment]
@WickedShell it will throw off a few things like filtering, but that is better than not running scheduling tasks. In practice you can be quite a long way off and it works

**Pull request merged**